refactor: more robust tagging index sync as recipient

benesjan opened the pull request:
Fixes #17775

In this PR I implement a new log sync algorithm we've come up with in Buenos Aires that should be fully robust against any kind of private log losses. The algorithm is explained in the body of `loadPrivateLogsForSenderRecipientPair` function. That function should be the entrypoint when reviewing this PR (if checking the utility functions first you would not have enough context).

Unfortunately this PR introduces a performance regression that is tracked by [this issue](https://linear.app/aztec-labs/issue/F-229/improve-log-sync-performance). I am fairly certain that the regression is caused by an unreasonable number of requests performed by the `loadLogsForRange` function - that will be tackled in a followup PR.

In this PR the algorithm is not yet integrated into the system. That is done in a [PR up the stack](https://github.com/AztecProtocol/aztec-packages/pull/19125).

The directory structure is not yet final - I wanted to keep this PR contained in one place to not have conflicts with Martin's PR. So please ignore that for now (will move stuff around in a final pass).

My plan is as follows:
1. Merge this PR,
2. fix the regression by modifying the `getLogsByTag` API such that it gives me all the info I need (now it doesn't give me block timestamp),
3. once that PR is merged most likely wait for Martin's refactor to be merged and then rebase and polish my integrating new log sync PR,
4. move some files around to make it all cuter.

## Update on regression of time in e2e tests
Unfortunately realized that this has caused a regression and log sync is now significantly slower.

Did 2 runs of the `e2e_l1_with_wall_time` test and on `next` the results are 337 and 334 seconds but with the new code it's 661 and 658 seconds.

Will work on dropping that down before merging that "integrating new log sync" PR.

benesjan commented:
* **#19290** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19290?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19289** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19289?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19233** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19233?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19125** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19125?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19161** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19161?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19158** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19158?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19157** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19157?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19030** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19030?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19030?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#17611** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17611?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17568** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17568?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17534** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17534?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17445** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17445?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17451** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17451?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on yarn-project/pxe/src/tagging/recipient_sync/utils/load_logs_for_range.ts:
```diff
@@ -0,0 +1,78 @@
+ * Gets private logs with their corresponding block timestamps and tagging indexes for the given index range, `app` and
+ * `secret`. At most load logs from blocks up to and including `anchorBlockNumber`.
+ *
+ * TODO: Optimize Aztec Node API such that this function performs only a single call.
```
This is the slow unreasonable function that performs a ton of requests to obtain the block timestamp.

Most of this will be deleted once `getLogsByTags` returns the block timestamp.

benesjan commented on yarn-project/pxe/src/tagging/recipient_sync/utils/load_logs_for_range.ts:
```diff
@@ -0,0 +1,78 @@
+    return [];
+  }
+
+  // Get unique block hashes
+  const uniqueBlockHashes = Array.from(new Set(privateLogsWithIndexes.map(({ log }) => log.blockHash.toBigInt()))).map(
+    hash => new Fr(hash),
+  );
+
+  // Get block headers for all unique block hashes
+  const blockHeaders = await Promise.all(uniqueBlockHashes.map(blockHash => aztecNode.getBlockHeaderByHash(blockHash)));
+
+  // Return logs with their corresponding block timestamps and tagging indexes
+  const result: Array<{ log: TxScopedL2Log; blockTimestamp: bigint; taggingIndex: number }> = [];
+  for (const { log, taggingIndex } of privateLogsWithIndexes) {
+    // TODO: Unify types of blockHash on log and on block header so we don't need to do this ugly conversion.
+    const logBlockHash = log.blockHash.toBigInt();
+    const logBlockHeader = blockHeaders[uniqueBlockHashes.findIndex(hash => hash.toBigInt() === logBlockHash)];
+    if (!logBlockHeader) {
+      // If the block header for a log cannot be found, it indicates a reorg occurred between `getLogsByTags` and
+      // `getBlockHeaderByHash`. It is correct and safe to ignore such logs because they have been pruned from
+      // the chain. PXE block synchronizer will reset any state following the reorg block.
+      continue;
+    }
+
+    result.push({ log, blockTimestamp: logBlockHeader.globalVariables.timestamp, taggingIndex });
+  }
+
+  return result;
```
All this will go away once `getLogsByTags` is optimized.

benesjan commented on yarn-project/pxe/src/tagging/recipient_sync/utils/load_logs_for_range.ts:
```diff
@@ -0,0 +1,78 @@
+    const logBlockHash = log.blockHash.toBigInt();
+    const logBlockHeader = blockHeaders[uniqueBlockHashes.findIndex(hash => hash.toBigInt() === logBlockHash)];
+    if (!logBlockHeader) {
+      // If the block header for a log cannot be found, it indicates a reorg occurred between `getLogsByTags` and
+      // `getBlockHeaderByHash`. It is correct and safe to ignore such logs because they have been pruned from
+      // the chain. PXE block synchronizer will reset any state following the reorg block.
+      continue;
```
I originally threw here an error but then it was difficult to update our reorg tests. I think just ignoring here the logs is correct. Also this code will go away soon.

mverzilli commented on yarn-project/pxe/src/tagging/recipient_sync/utils/load_logs_for_range.ts:
```diff
@@ -0,0 +1,78 @@
+  );
+
+  // Get block headers for all unique block hashes
+  const blockHeaders = await Promise.all(uniqueBlockHashes.map(blockHash => aztecNode.getBlockHeaderByHash(blockHash)));
```
Definitely not for this PR, but eventually it would be nice to add an endpoint to the node where we can solve this in one call

benesjan replied:
Already done in https://github.com/AztecProtocol/aztec-packages/pull/19157 😊

mverzilli commented on yarn-project/pxe/src/tagging/recipient_sync/utils/load_logs_for_range.ts:
```diff
@@ -0,0 +1,78 @@
+    const logBlockHash = log.blockHash.toBigInt();
+    const logBlockHeader = blockHeaders[uniqueBlockHashes.findIndex(hash => hash.toBigInt() === logBlockHash)];
+    if (!logBlockHeader) {
+      // If the block header for a log cannot be found, it indicates a reorg occurred between `getLogsByTags` and
+      // `getBlockHeaderByHash`. It is correct and safe to ignore such logs because they have been pruned from
+      // the chain. PXE block synchronizer will reset any state following the reorg block.
```
cool

mverzilli commented on yarn-project/pxe/src/tagging/recipient_sync/utils/load_logs_for_range.ts:
```diff
@@ -0,0 +1,78 @@
+import type { BlockNumber } from '@aztec/foundation/branded-types';
+import { Fr } from '@aztec/foundation/curves/bn254';
+import type { AztecAddress } from '@aztec/stdlib/aztec-address';
+import type { AztecNode } from '@aztec/stdlib/interfaces/server';
```
based on recent experience, you might want this import to point to /client instead of server to avoid big bundles

benesjan replied:
Good point!

https://github.com/AztecProtocol/aztec-packages/pull/19030/changes/053f6ac6b8c37a29c86c4319ca4ca16c002ee196

mverzilli commented on yarn-project/pxe/src/tagging/recipient_sync/utils/load_logs_for_range.test.ts:
```diff
@@ -0,0 +1,355 @@
+
+  beforeAll(async () => {
+    secret = DirectionalAppTaggingSecret.fromString(Fr.random().toString());
+    app = await AztecAddress.random();
```
I think in most other places we call this `contractAddress`, do we want to adopt a new convention here?

benesjan replied:
As discussed in scrum would ignore this for now.

mverzilli commented on yarn-project/pxe/src/tagging/recipient_sync/utils/load_logs_for_range.ts:
```diff
@@ -0,0 +1,78 @@
+  secret: DirectionalAppTaggingSecret,
+  app: AztecAddress,
+  aztecNode: AztecNode,
+  start: number,
+  end: number,
```
would be nice to get some doc comments further explaining these params, and particularly clarifying whether the bounds are inclusive or not

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/19030/commits/a4fc91741bd2fc69b69564a217436e38c88380f6

mverzilli commented on yarn-project/pxe/src/tagging/recipient_sync/new_recipient_tagging_data_provider.ts:
```diff
@@ -0,0 +1,50 @@
+export class NewRecipientTaggingDataProvider {
+  #store: AztecAsyncKVStore;
+
+  #highestAgedIndex: AztecAsyncMap<string, number>;
+  #highestFinalizedIndex: AztecAsyncMap<string, number>;
```
Do we need two different maps for this? I was expecting this to be a single map like `secret => {highestAgedIndex, highestFinalizedIndex}`

benesjan replied:
I think it's easier implementation-wise to have 2 maps because we are setting and obtaining these individually.

mverzilli replied:
I'm honestly not sold that just because two fields are read/write in different places they should be persisted in different structures... 

I'm thinking a little bit ahead, for DB integrity we will likely need to store versions of these changes in memory and it might become somewhat noisier. But I can deal with that later if that becomes an issue, so no prob with moving forward with this as is

benesjan replied:
> for DB integrity we will likely need to store versions of these changes in memory

I don't see how having a single map there would affect it as it's an implementation detail of the provider class that is not exposed to the outside world.

Could you expand?

But anyway this is so simple that I think it's ok to just do it when needed.

mverzilli commented on yarn-project/pxe/src/tagging/recipient_sync/new_recipient_tagging_data_provider.ts:
```diff
@@ -0,0 +1,50 @@
+ *
+ * TODO(benesjan): Rename as to RecipientTaggingDataProvider and relocate once the old tagging sync is purged.
+ */
+export class NewRecipientTaggingDataProvider {
```
What happens when there's reorgs? I was expecting some sort of pruning to happen 🤔

benesjan replied:
That's a nice feature of this algo - both aged and finalized indexes correspond to a finalized block and hence by definition cannot be affected by reorgs.

Added there a comment on this: https://github.com/AztecProtocol/aztec-packages/pull/19030/commits/827066e00f3acea47f747f3ee8cc022f0f75b966

mverzilli replied:
Ahhhh right makes a lot of sense and is cute

mverzilli commented on yarn-project/pxe/src/tagging/recipient_sync/load_private_logs_for_sender_recipient_pair.ts:
```diff
@@ -0,0 +1,130 @@
+import type { DirectionalAppTaggingSecret, TxScopedL2Log } from '@aztec/stdlib/logs';
+
+// TODO(benesjan): Rename WINDOW_LEN to something more descriptive.
+import { WINDOW_LEN } from '../sync/sync_sender_tagging_indexes.js';
```
`UNFINALIZED_TAGGING_INDEXES_WINDOW_LEN`?

mverzilli replied:
`TAGGING_INDEX_RANGE_SIZE`?

benesjan replied:
I like `UNFINALIZED_TAGGING_INDEXES_WINDOW_LEN` 👍 

Addressed in https://github.com/AztecProtocol/aztec-packages/pull/19030/commits/c1b2cc6834f4b47eee5bc491599e0b0d27f7618a

mverzilli commented on yarn-project/pxe/src/tagging/recipient_sync/load_private_logs_for_sender_recipient_pair.ts:
```diff
@@ -0,0 +1,130 @@
+ * Loads private logs for `app` and  sender-recipient pair defined by `secret` and updates the highest aged and
+ * finalized indexes in the db. At most load logs from blocks up to and including `anchorBlockNumber`.
+ *
+ * @dev This function can be safely executed "in parallel" for other sender-recipient pairs because the data in
```
cool

mverzilli commented on yarn-project/pxe/src/tagging/recipient_sync/load_private_logs_for_sender_recipient_pair.ts:
```diff
@@ -0,0 +1,130 @@
+  // When a sender chooses a tagging index, they will select an index that is at most `WINDOW_LEN` greater than
+  // the highest finalized index. If that index was already used, they will throw an error. For this reason we
+  // don't have to look further than `highestFinalizedIndex + WINDOW_LEN`.
+
```
excellent explanation 👍

mverzilli commented on yarn-project/pxe/src/tagging/recipient_sync/load_private_logs_for_sender_recipient_pair.ts:
```diff
@@ -0,0 +1,130 @@
+  // the highest finalized index. If that index was already used, they will throw an error. For this reason we
+  // don't have to look further than `highestFinalizedIndex + WINDOW_LEN`.
+
+  let finalizedBlockNumber: number, currentTimestamp: bigint;
```
nitpick: strange that we don't have any lint rules about this, but I'd prefer two lines each with their own let expression

mverzilli replied:
ahhh never mind, you're assigning through the block below, interesting, never used that idiom

benesjan replied:
Just wanted to communicate which variables are relevant for the following code and which are not.

mverzilli reviewed (changes requested):
Looks awesome, just want to sync particularly regarding data storage

mverzilli reviewed (approved):
🚀

benesjan committed:
refactor: more robust tagging index sync as recipient

Fixes #17775

In this PR I implement a new log sync algorithm we've come up with in Buenos Aires that should be fully robust against any kind of private log losses. The algorithm is explained in the body of `loadPrivateLogsForSenderRecipientPair` function. That function should be the entrypoint when reviewing this PR (if checking the utility functions first you would not have enough context).

Unfortunately this PR introduces a performance regression that is tracked by [this issue](https://linear.app/aztec-labs/issue/F-229/improve-log-sync-performance). I am fairly certain that the regression is caused by an unreasonable number of requests performed by the `loadLogsForRange` function - that will be tackled in a followup PR.

In this PR the algorithm is not yet integrated into the system. That is done in a [PR up the stack](https://github.com/AztecProtocol/aztec-packages/pull/19125).

The directory structure is not yet final - I wanted to keep this PR contained in one place to not have conflicts with Martin's PR. So please ignore that for now (will move stuff around in a final pass).

My plan is as follows:
1. Merge this PR,
2. fix the regression by modifying the `getLogsByTag` API such that it gives me all the info I need (now it doesn't give me block timestamp),
3. once that PR is merged most likely wait for Martin's refactor to be merged and then rebase and polish my integrating new log sync PR,
4. move some files around to make it all cuter.

## Update on regression of time in e2e tests
Unfortunately realized that this has caused a regression and log sync is now significantly slower.

Did 2 runs of the `e2e_l1_with_wall_time` test and on `next` the results are 337 and 334 seconds but with the new code it's 661 and 658 seconds.

Will work on dropping that down before merging that "integrating new log sync" PR.
